### PR TITLE
[MISC] rename metafunction to type trait.

### LINKED ISSUE #243: metafunction → type_traits
We should switch the vocabulary of the standard also for metafunctions, this means:

- [ ] Use new names:
  * "value metafunction" → TypeTrait
  * "type metafunction" → TransformationTrait
- [ ] update API documentation
- [ ] formally introduce concepts unary_type_trait_concept and binary_type_trait_concept fpr TypeTraits and transformation_trait_concept for TransformationTrait; TypeTraits shall inherit std::integral_constant
- [ ] change "metafunction" wiki article to "Type traits, traits-types and metafunctions" and explain all

### PR DESCRIPTION
**Pull request by smehringer**
Resolves #243

### CODE REVIEW

File: include/seqan3/core/type_traits/basic.hpp [hunk 1]
```diff
@@ -60,14 +61,14 @@ using remove_rvalue_reference_t = typename remove_rvalue_reference<t>::type;
...
-/*!\brief Whether a type std::is_default_constructible in `constexpr`-context [UnaryTypeTrait].
+/*!\brief Whether a type std::is_default_constructible in `constexpr`-context [TypeTrait].
  * \tparam t The type to operate on.
  */
 template <typename t>
 struct is_constexpr_default_constructible : std::false_type
 {};
 
-/*!\brief Whether a type std::is_default_constructible in `constexpr`-context [UnaryTypeTrait specialisation].
+/*!\brief Whether a type std::is_default_constructible in `constexpr`-context [TypeTrait specialisation].
```
Comment (marehr) [inline include/seqan3/core/type_traits/basic.hpp]: We have no `\implements` / concept for this?

File: include/seqan3/alphabet/composite/alphabet_tuple_base.hpp [hunk 2]
```diff
@@ -687,7 +687,8 @@ constexpr bool operator>=(indirect_component_type const lhs,
 namespace std
 {
 
-//!\brief Obtains the type of the specified element.
+//!\brief Obtains the type of the specified element .
```
Comment (marehr) [inline include/seqan3/alphabet/composite/alphabet_tuple_base.hpp]: Space between dot.

File: include/seqan3/core/pod_tuple.hpp [hunk 3]
```diff
@@ -351,8 +352,10 @@ struct tuple_element<i, t<types...>>
     using type = meta::at_c<meta::list<types...>, i>;
 };
 
-//!\brief Provides access to the number of elements in a tuple as a compile-time constant expression.
-//!\relates seqan3::pod_tuple
+/*!\brief Provides access to the number of elements in a tuple as a compile-time constant expression
+ *!       [UnaryTypeTrait specialisation].
```
Comment (marehr) [inline include/seqan3/core/pod_tuple.hpp]:
```suggestion
 *        [UnaryTypeTrait specialisation].
```

File: include/seqan3/core/pod_tuple.hpp [hunk 4]
```diff
@@ -351,8 +352,10 @@ struct tuple_element<i, t<types...>>
...
+ *!\relates seqan3::pod_tuple
```
Comment (marehr) [inline include/seqan3/core/pod_tuple.hpp]:
```suggestion
 * \relates seqan3::pod_tuple
```

File: include/seqan3/core/type_traits/all.hpp [hunk 5]
```diff
@@ -6,21 +6,21 @@
...
-#include <seqan3/core/metafunction/basic.hpp>
-#include <seqan3/core/metafunction/deferred_crtp_base.hpp>
-#include <seqan3/core/metafunction/iterator.hpp>
-#include <seqan3/core/metafunction/range.hpp>
-#include <seqan3/core/metafunction/template_inspection.hpp>
-#include <seqan3/core/metafunction/transformation_trait_or.hpp>
-
-/*!\defgroup metafunction Metafunction
- * \brief Provide various metafunctions.
+/*!\defgroup type_traits Type Traits
```
Comment (marehr) [inline include/seqan3/core/type_traits/all.hpp]: I think most TypeTraits don't use this group.

File: include/seqan3/core/type_traits/basic.hpp (hunk 1 above)
Comment (marehr) [reply include/seqan3/core/type_traits/basic.hpp]: What exactly is a TypeTrait? There is no definition at cppreference: https://en.cppreference.com/w/cpp/named_req/TypeTrait.

File: include/seqan3/core/type_traits/function.hpp [hunk 6]
```diff
@@ -6,7 +6,7 @@
 // -----------------------------------------------------------------------------------------------------
 
 /*!\file
- * \brief Provides various metafunctions for use on functions.
+ * \brief Provides various type traits for use on functions.
```
Comment (marehr) [inline include/seqan3/core/type_traits/function.hpp]: > various

namely one xD

File: include/seqan3/io/structure_file/input.hpp [hunk 7]
```diff
@@ -1151,8 +1150,8 @@ struct tuple_size<seqan3::structure_file_in<traits_type, selected_field_ids, val
     static constexpr size_t value = selected_field_ids::as_array.size();
 };
 
-/*!\brief std::tuple_element overload for column-like access.
- * [metafunction specialisation for seqan3::structure_file_in]
+/*!\brief std::tuple_element overload for column-like access .
```
Comment (marehr) [inline include/seqan3/io/structure_file/input.hpp]: space between dot

File: include/seqan3/core/type_traits/iterator.hpp [hunk 8]
```diff
@@ -16,22 +16,23 @@
...
  * \{
  */
 
 // ----------------------------------------------------------------------------
 // value_type
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `value_type` of another type [specialisation for input iterators].
- * \tparam it_t The type you wish to query; must satisfy std::input_Iterator.
+/*!\brief Exposes the `value_type` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/iterator.hpp]: space between the dot

File: include/seqan3/core/type_traits/pre.hpp [hunk 9]
```diff
@@ -17,167 +17,173 @@
...
-// see specialisation for ranges in core/metafunction/range.hpp
+// see specialisation for iterators in core/type_traits/iterator.hpp
+// see specialisation for ranges in core/type_traits/range.hpp
 
 // ----------------------------------------------------------------------------
 // reference
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `reference` of another type [metafunction declaration].
+/*!\brief Exposes the `reference` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/pre.hpp]: space between the dot

File: include/seqan3/core/type_traits/pre.hpp [hunk 10]
```diff
@@ -17,167 +17,173 @@
...
-// see specialisation for ranges in core/metafunction/range.hpp
+// see specialisation for iterators in core/type_traits/iterator.hpp
+// see specialisation for ranges in core/type_traits/range.hpp
 
 // ----------------------------------------------------------------------------
 // const_reference
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `const_reference` of another type [metafunction declaration].
+/*!\brief Exposes the `const_reference` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/pre.hpp]: space between the dot

File: include/seqan3/core/type_traits/pre.hpp [hunk 11]
```diff
@@ -17,167 +17,173 @@
...
 // no specialisation for iterators
-// see specialisation for ranges in core/metafunction/range.hpp
+// see specialisation for ranges in core/type_traits/range.hpp
 
 // ----------------------------------------------------------------------------
 // difference_type
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `difference_type` of another type [metafunction declaration].
+/*!\brief Exposes the `difference_type` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/pre.hpp]: space between the dot

File: include/seqan3/core/type_traits/deferred_crtp_base.hpp [hunk 12]
```diff
@@ -47,8 +48,9 @@ struct deferred_crtp_base
     using invoke = crtp_base<derived_t, args_t...>;
 };
 
-/*!\brief Template alias to instantiate the deferred crtp base with the derived class.
- * \ingroup metafunction
+/*!\brief Template alias to instantiate the deferred crtp base with the derived class .
```
Comment (marehr) [inline include/seqan3/core/type_traits/deferred_crtp_base.hpp]: space between the dot

File: include/seqan3/core/type_traits/pre.hpp [hunk 13]
```diff
@@ -17,167 +17,173 @@
...
-// see specialisation for ranges in core/metafunction/range.hpp
+// see specialisation for iterators in core/type_traits/iterator.hpp
+// see specialisation for ranges in core/type_traits/range.hpp
 
 // ----------------------------------------------------------------------------
 // size_type
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `size_type` of another type [metafunction declaration].
+/*!\brief Exposes the `size_type` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/pre.hpp]: space between the dot

File: include/seqan3/core/type_traits/range.hpp [hunk 14]
```diff
@@ -40,16 +40,17 @@ SEQAN3_CONCEPT has_value_type = requires { typename value_type_t<remove_cvref_t<
...
  * \{
  */
 
 // ----------------------------------------------------------------------------
 // value_type
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `value_type` of another type [specialisation for input ranges].
- * \tparam t The type you wish to query; must satisfy std::ranges::InputRange.
+/*!\brief Exposes the `value_type` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/range.hpp]: space between the dot

File: include/seqan3/core/type_traits/range.hpp [hunk 15]
```diff
@@ -65,8 +66,9 @@ struct value_type<rng_t>
 // reference
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `reference` of another type [specialisation for input ranges].
- * \tparam t The type you wish to query; must satisfy std::ranges::InputRange.
+/*!\brief Exposes the `reference` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/range.hpp]: space between the dot

File: include/seqan3/core/type_traits/range.hpp [hunk 16]
```diff
@@ -82,8 +84,9 @@ struct reference<rng_t>
 // rvalue_reference
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `rvalue_reference` of another type [specialisation for input ranges].
- * \tparam t The type you wish to query; must satisfy std::ranges::InputRange.
+/*!\brief Exposes the `rvalue_reference` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/range.hpp]: space between the dot

File: include/seqan3/core/type_traits/range.hpp [hunk 17]
```diff
@@ -99,8 +102,9 @@ struct rvalue_reference<rng_t>
 // const_reference
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `const_reference` of another type [specialisation for input ranges].
- * \tparam t The type you wish to query; must satisfy std::ranges::InputRange.
+/*!\brief Exposes the `const_reference` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/range.hpp]: space between the dot

File: include/seqan3/core/type_traits/range.hpp [hunk 18]
```diff
@@ -116,8 +120,9 @@ struct const_reference<rng_t>
 // difference_type
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `difference_type` of another type [specialisation for ranges].
- * \tparam t The type you wish to query; must satisfy std::ranges::InputRange.
+/*!\brief Exposes the `difference_type` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/range.hpp]: space between the dot

File: include/seqan3/core/type_traits/range.hpp [hunk 19]
```diff
@@ -133,8 +138,9 @@ struct difference_type<rng_t>
 // size_type
 // ----------------------------------------------------------------------------
 
-/*!\brief Type metafunction that returns the `size_type` of another type [specialisation for sized ranges].
- * \tparam t The type you wish to query; must satisfy std::ranges::SizedRange.
+/*!\brief Exposes the `size_type` of another type .
```
Comment (marehr) [inline include/seqan3/core/type_traits/range.hpp]: space between the dot

File: include/seqan3/core/type_traits/range.hpp [hunk 20]
```diff
@@ -152,12 +158,13 @@ struct size_type<rng_t>
 
 //NOTE(h-2): this could be moved to a separate file, because it also applies to iterators
 
-/*!\brief Recursively determines the `value_type` on containers and/or iterators [Type metafunction].
- * \tparam t The type to recurse on; must have `std::ranges::value_type_t<rng_t>`
+/*!\brief Recursively determines the `value_type` on containers and/or iterators .
```
Comment (marehr) [inline include/seqan3/core/type_traits/range.hpp]: space between the dot

File: include/seqan3/core/type_traits/transformation_trait_or.hpp [hunk 21]
```diff
@@ -22,16 +22,17 @@
 namespace seqan3::detail
 {
 
-/*!\brief This gives a fallback type if *type_t::type* is not defined.
- * \ingroup metafunction
+/*!\brief This gives a fallback type if *type_t::type* is not defined .
```
Comment (marehr) [inline include/seqan3/core/type_traits/transformation_trait_or.hpp]: space between the dot

File: include/seqan3/io/alignment_file/input.hpp [hunk 22]
```diff
@@ -1136,7 +1136,8 @@ struct tuple_size<seqan3::alignment_file_input<traits_type, selected_field_ids,
     static constexpr size_t value = selected_field_ids::as_array.size();
 };
 
-//!\brief std::tuple_element overload for column-like access. [metafunction specialisation for seqan3::alignment_file_input]
+//!\brief std::tuple_element overload for column-like access .
```
Comment (marehr) [inline include/seqan3/io/alignment_file/input.hpp]: space between the dot

File: include/seqan3/io/detail/record.hpp [hunk 23]
```diff
@@ -38,7 +38,8 @@ SEQAN3_CONCEPT Fields = is_value_specialisation_of_v<t, fields>;
 // select_types_with_ids
 // ----------------------------------------------------------------------------
 
-/*!\brief A type metafunction that selects a subset of types based on their IDs.
+/*!\brief Exposes a subset of types as a seqan3::type_list selected based on their IDs .
```
Comment (marehr) [inline include/seqan3/io/detail/record.hpp]: space between the dot

File: include/seqan3/io/record.hpp [hunk 24]
```diff
@@ -209,7 +209,8 @@ struct tuple_size<seqan3::record<field_types, field_ids>>
     static constexpr size_t value = tuple_size_v<typename seqan3::record<field_types, field_ids>::base_type>;
 };
 
-/*!\brief Value metafunction specialisation for seqan3::record; returns the type of an element in the record.
+/*!\brief Exposes the type of an element in the record .
```
Comment (marehr) [inline include/seqan3/io/record.hpp]: space between the dot

File: include/seqan3/io/sequence_file/input.hpp [hunk 25]
```diff
@@ -858,7 +858,8 @@ struct tuple_size<seqan3::sequence_file_input<traits_type, selected_field_ids, v
     static constexpr size_t value = selected_field_ids::as_array.size();
 };
 
-//!\brief std::tuple_element overload for column-like access. [metafunction specialisation for seqan3::sequence_file_input]
+//!\brief std::tuple_element overload for column-like access .
```
Comment (marehr) [inline include/seqan3/io/sequence_file/input.hpp]: space between the dot

File: include/seqan3/alignment/pairwise/align_result_selector.hpp [hunk 26]
```diff
@@ -18,15 +18,16 @@
...
+#include <seqan3/core/type_traits/basic.hpp>
+#include <seqan3/core/type_traits/range.hpp>
 #include <seqan3/core/type_list.hpp>
 #include <seqan3/std/ranges>
 
 namespace seqan3::detail
 {
 
-/*!\brief Helper metafunction to select the alignment result type based on the configuration.
+/*!\brief Exposes the alignment result type based on the configuration .
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/alignment/pairwise/align_result_selector.hpp]: Ok.. - two "space between the dot" and please check the rest - 
would have been sufficient as this clearly is the result of a regex..

**Comment by h-2** [comment]
Ok so I think in general it's good to avoid having to write the same three words after every line, but I don't really see how 
`\this_is_a_type_trait FOOBAR` is that much better than `\brief FOOBAR [this is a type trait]` and I am reluctant to start a huge discussion about something as simple as a rename operation.

Since the type trait, its specialisations and the shortcut all appear as distinct entities in doxygen, I thought that it would be good to distinguish them be using the respective keyword in the brief. I am totally open to using different keywords or () instead of [] or whatever, but I haven't seen a good arguments for either, yet.

**Comment by rrahn** [comment]
In general, I don't care if there is a macro that generates the docs for it or not. I would slightly prefer the macro version in order get a consistent documentation, but having a wiki entry which describes the format (so I can use it as a snippet) is just fine for me. More than the how we document it, it is I think it is more important to get the naming right. There are UnaryTypeTraits, BinaryTypeTraits and TransformationTraits. So if we touch and move all the stuff anyway, that could be fixed right along. Morover, the shortcut is described as Helper template or Helper variable template throughout  cppref and we should stick to this. It would be cool if we could include code from code. Like define a doxygen snippet around the helper template and use this snippet to add it in the documentation of the respective trait definition. But I guess this is far fetched for our beloved doxygen?

**Comment by smehringer** [comment]
> I think it is more important to get the naming right. There are UnaryTypeTraits, BinaryTypeTraits and TransformationTraits (if not done already - haven't looked into the changes)

I already started tackling this while changing the naming but I will only do this for stuff that I touch anyway or come across by chance. To go through the whole documentation and check each metafunction manually can be part of a later PR (if I have time I will included it here but I cannot promise).

> Like define a doxygen snippet around the helper template and use this snippet to add it in the documentation of the respective trait definition.

I don't see how this is done with doxygen.

File: include/seqan3/alphabet/adaptation/char.hpp [hunk 27]
```diff
@@ -35,7 +35,7 @@ namespace seqan3::detail
 //!\addtogroup adaptation
 //!\{
 
-//!\brief Whether a type is `char`, `char16_t`, `char32_t` or `wchar_t`.
+//!\brief Whether a type is `char`, `char16_t`, `char32_t` or `wchar_t` [TypeTrait shortcut].
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/alphabet/adaptation/char.hpp]: I removed [] in favour of ()

I removed the "shortcut"

File: include/seqan3/core/type_traits/basic.hpp [hunk 28]
```diff
@@ -90,16 +91,16 @@ inline constexpr bool is_constexpr_default_constructible_v = is_constexpr_defaul
...
+/*!\addtogroup type_traits
  * \{
  */
 
 // ----------------------------------------------------------------------------
 // deferred_type
 // ----------------------------------------------------------------------------
 
-/*!\brief Return the type identity; further arguments are ignored, but can make this type dependent if they are
- *        [TransformationTrait].
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/core/type_traits/basic.hpp]: I remove it, if it actually implements the `TransformationTrait` as the information is redundant then

File: include/seqan3/core/type_traits/function.hpp (hunk 6 above)
Comment (smehringer) [reply include/seqan3/core/type_traits/function.hpp]: well maybe there are more in the future

File: include/seqan3/core/type_traits/basic.hpp [hunk 29]
```diff
@@ -15,20 +15,20 @@
...
+/*!\addtogroup type_traits
  * \{
  */
 
 // ----------------------------------------------------------------------------
 // remove_cvref_t
 // ----------------------------------------------------------------------------
 
-/*!\brief Return the input type with `const`, `volatile` and references removed [Type metafunction].
+/*!\brief Return the input type with `const`, `volatile` and references removed (type trait).
```
Comment (marehr) [inline include/seqan3/core/type_traits/basic.hpp]: here lower case

File: include/seqan3/core/type_traits/basic.hpp [hunk 30]
```diff
@@ -77,9 +80,9 @@ template <typename t>
 struct is_constexpr_default_constructible<t> : std::integral_constant<bool, SEQAN3_IS_CONSTEXPR(t{})>
 {};
 
-/*!\brief Whether a type std::is_default_constructible in `constexpr`-context [UnaryTypeTrait shortcut].
+/*!\brief Whether a type std::is_default_constructible in `constexpr`-context (TypeTrait shortcut).
```
Comment (marehr) [inline include/seqan3/core/type_traits/basic.hpp]: here camel case

File: include/seqan3/core/type_traits/range.hpp [hunk 31]
```diff
@@ -178,22 +185,29 @@ struct innermost_value_type<t>
...
+//!\brief Shortcut for seqan3::innermost_value_type (TransformationTrait shortcut).
+//!\see seqan3::innermost_value_type
 template <typename t>
 using innermost_value_type_t = typename innermost_value_type<t>::type;
 
+//!\}
+
+/*!\addtogroup type_traits
+ * \{
+ */
```
Comment (marehr) [inline include/seqan3/core/type_traits/range.hpp]: dupl :)

File: test/unit/core/type_traits/template_inspection_test.cpp [hunk 32]
```diff
@@ -9,11 +9,22 @@
...
 #include <gtest/gtest.h>
 
-#include <seqan3/core/metafunction/template_inspection.hpp>
+#include <seqan3/core/type_traits/concept.hpp>
+#include <seqan3/core/type_traits/template_inspection.hpp>
 #include <seqan3/core/type_list.hpp>
 
 using namespace seqan3;
 
+TEST(template_inspect, concept)
```
Comment (marehr) [inline test/unit/core/type_traits/template_inspection_test.cpp]: we name the test `concept_check` through out the library

File: include/seqan3/core/type_traits/basic.hpp (hunk 29 above)
Comment (smehringer) [reply include/seqan3/core/type_traits/basic.hpp]: yes it should be lower case unless it describes a concept. good catch

File: test/unit/core/type_traits/template_inspection_test.cpp (hunk 32 above)
Comment (smehringer) [reply test/unit/core/type_traits/template_inspection_test.cpp]: well not everywhere but in most places. I changed it accordingly

**Pull request merged**